File: apps/Lodgr.Api/Features/Rooms/Controllers/RoomsController.cs
using Lodgr.Api.Common.Pagination;
using Microsoft.AspNetCore.Mvc;

namespace Lodgr.Api.Features.Rooms;

[ApiController]
[Route("api/rooms")]
public class RoomsController(IRoomService service) : ControllerBase
{
    [HttpGet]
    [ProducesResponseType(typeof(PagedResult<RoomListItemResponse>), StatusCodes.Status200OK)]
    public async Task<ActionResult<PagedResult<RoomListItemResponse>>> GetRooms(
        [FromQuery] GetRoomsRequest request,
        CancellationToken cancellationToken)
    {
        var result = await service.GetPagedAsync(request, cancellationToken);
        return Ok(result);
    }

    [HttpGet("{roomId:long}")]
    [ProducesResponseType(typeof(RoomDetailResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<ActionResult<RoomDetailResponse>> GetRoomById(
        [FromRoute] long roomId,
        [FromQuery] bool includeDeleted,
        CancellationToken cancellationToken)
    {
        var result = await service.GetByIdAsync(roomId, includeDeleted, cancellationToken);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpGet("{roomId:long}/contracts")]
    [ProducesResponseType(typeof(PagedResult<RoomContractListItemResponse>), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<ActionResult<PagedResult<RoomContractListItemResponse>>> GetRoomContracts(
        [FromRoute] long roomId,
        [FromQuery] GetRoomContractsRequest request,
        CancellationToken cancellationToken)
    {
        var result = await service.GetPagedContractsAsync(roomId, request, cancellationToken);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpPost]
    [ProducesResponseType(typeof(RoomDetailResponse), StatusCodes.Status201Created)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    public async Task<ActionResult<RoomDetailResponse>> CreateRoom(
        [FromBody] CreateRoomRequest request,
        CancellationToken cancellationToken)
    {
        var result = await service.CreateAsync(request, cancellationToken);
        return CreatedAtAction(nameof(GetRoomById), new { roomId = result.RoomId }, result);
    }

    [HttpPut("{roomId:long}")]
    [ProducesResponseType(typeof(RoomDetailResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    public async Task<ActionResult<RoomDetailResponse>> UpdateRoom(
        [FromRoute] long roomId,
        [FromBody] UpdateRoomRequest request,
        CancellationToken cancellationToken)
    {
        var result = await service.UpdateAsync(roomId, request, cancellationToken);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpDelete("{roomId:long}")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> DeleteRoom(
        [FromRoute] long roomId,
        CancellationToken cancellationToken)
    {
        var deleted = await service.SoftDeleteAsync(roomId, cancellationToken);
        return deleted ? NoContent() : NotFound();
    }
}

File: apps/api/src/main/java/org/lodgr/api/features/rooms/RoomsController.java
package org.lodgr.api.features.rooms;

import static org.lodgr.api.features.rooms.RoomDtos.*;

import jakarta.validation.Valid;
import org.lodgr.api.pagination.PagedResult;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@Validated
@RequestMapping("/api/rooms")
public class RoomsController {
    private final RoomService roomService;

    public RoomsController(RoomService roomService) {
        this.roomService = roomService;
    }

    @GetMapping
    public ResponseEntity<PagedResult<RoomListItemResponse>> getRooms(
            @RequestParam(required = false) Integer pageNumber,
            @RequestParam(required = false) Integer pageSize,
            @RequestParam(required = false) Long buildingId,
            @RequestParam(required = false, defaultValue = "false") Boolean includeDeleted,
            @RequestParam(required = false) String keyword,
            @RequestParam(required = false) String operationalStatus) {
        GetRoomsQuery query = new GetRoomsQuery(pageNumber, pageSize, buildingId, includeDeleted, keyword,
                operationalStatus);
        return ResponseEntity.ok(roomService.getPaged(query));
    }

    @GetMapping("/{roomId}")
    public ResponseEntity<RoomDetailResponse> getRoomById(
            @PathVariable Long roomId,
            @RequestParam(required = false, defaultValue = "false") boolean includeDeleted) {
        RoomDetailResponse result = roomService.getById(roomId, includeDeleted);
        if (result == null) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.ok(result);
    }

    @GetMapping("/{roomId}/contracts")
    public ResponseEntity<PagedResult<RoomContractListItemResponse>> getRoomContracts(
            @PathVariable Long roomId,
            @RequestParam(required = false) Integer pageNumber,
            @RequestParam(required = false) Integer pageSize,
            @RequestParam(required = false, defaultValue = "true") Boolean includeInactive) {
        GetRoomContractsQuery query = new GetRoomContractsQuery(pageNumber, pageSize, includeInactive);
        PagedResult<RoomContractListItemResponse> result = roomService.getPagedContracts(roomId, query);

        if (result == null) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.ok(result);
    }

    @PostMapping
    public ResponseEntity<RoomDetailResponse> createRoom(@Valid @RequestBody CreateRoomRequest request) {
        RoomDetailResponse result = roomService.create(request);
        return ResponseEntity.status(HttpStatus.CREATED).body(result);
    }

    @PutMapping("/{roomId}")
    public ResponseEntity<RoomDetailResponse> updateRoom(
            @PathVariable Long roomId,
            @Valid @RequestBody UpdateRoomRequest request) {
        RoomDetailResponse result = roomService.update(roomId, request);
        if (result == null) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.ok(result);
    }

    @DeleteMapping("/{roomId}")
    public ResponseEntity<Void> deleteRoom(@PathVariable Long roomId) {
        boolean deleted = roomService.softDelete(roomId);
        if (!deleted) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.noContent().build();
    }
}
